## Data and code supporting: "Impacts of Antiretroviral Therapy on the Oral Microbiome and Periodontal Health of Feline Immunodeficiency Virus Positive Cats"

### Laura Bashor a, Jennifer E. Rawlinson b, Christopher P. Kozakiewicz c,d,e, Elisa Behzadi a, Craig Miller f, Jeffrey Kim g, Megan Conry b, Mary Nehring a, Scott Carver h,i,j, Zaid Abdo a, Susan VandeWoude a

a Department of Microbiology, Immunology, and Pathology, Colorado State University, Fort Collins, CO, USA
b Department of Clinical Sciences, Colorado State University, Fort Collins, CO, USA
c W.K. Kellogg Biological Station, Michigan State University, Hickory Corners, MI, USA
d Department of Integrative Biology, Michigan State University, East Lansing, MI, USA
e Ecology, Evolution, and Behavior Program, Michigan State University, East Lansing, MI, USA
f Department of Veterinary Pathobiology, College of Veterinary Medicine, Oklahoma State University, Stillwater, OK, USA
g Comparative Medicine Research Unit, School of Medicine, University of Louisville, Louisville, KY, USA
h Odum School of Ecology, University of Georgia, Athens, GA
i Center for the Ecology of Infectious Diseases, University of Georgia, Athens, GA
j Department of Biological Sciences, University of Tasmania, Tasmania, Australia


Full manuscript available soon.

Raw 16S sequencing reads available in NCBI SRA BioProject PRJNA1076586

Next-generation sequencing datasets were subjected to quality control and trimming (Trimmomatic) prior to analysis with mothur software following a modified MiSeq SOP: https://mothur.org/wiki/miseq_sop/ (Schloss et al., 2009). For the mothur analysis, data were reduced to unique sequences and their associated frequencies. The unique sequences were then classified using the SILVA taxonomic database version 138.1 (Pruesse et al., 2007) obtained from https://mothur.org/wiki/silva_reference_files/. Chimeric sequences and other potential contaminants (such as mitochondria) were removed. Data were then clustered into operational taxonomic units (OTUs) using distance-based greedy clustering at 97% identity and reclassified to identify the prevalent microbial taxa and the associated abundances.
